File: backend/app/domain/schemas/paciente.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, Literal
from datetime import date
from .alergia import AlergiaRead
from .antecedente import AntecedenteRead
from .obra_social import ObraSocialRead
from .grupo_sanguineo import GrupoSanguineoRead

GrupoSanguineoType = Literal["A+", "A-", "B+", "B-", "AB+", "AB-", "0+", "0-"]

class PacienteBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    dni: int
    nombre: str
    apellido: str
    fecha_nacimiento: date
    email: str
    telefono: str

class PacienteCreate(PacienteBase):
    # ✅ EJEMPLO PARA SWAGGER POST
    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "dni": 12345678,
                "nombre": "Juan Carlos",
                "apellido": "Pérez García",
                "fecha_nacimiento": "1990-05-15",
                "email": "juan.perez@email.com",
                "telefono": "+54911234567",
                "nombre_grupo_sanguineo": "A+",
                "nombre_obra_social": "OSDE",
                "ids_alergias": [1, 2],
                "ids_antecedentes": [1]
            }
        }
    )
    
    ids_alergias: Optional[list[int]] = None
    ids_antecedentes: Optional[list[int]] = None
    nombre_obra_social: Optional[str] = None
    nombre_grupo_sanguineo: GrupoSanguineoType

class PacienteUpdate(BaseModel):
    # ✅ EJEMPLO PARA SWAGGER PUT
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "nombre": "Juan Carlos Actualizado",
                "apellido": "Pérez García",
                "fecha_nacimiento": "1990-05-15",
                "email": "juan.actualizado@email.com",
                "telefono": "+54911234567",
                "nombre_grupo_sanguineo": "B+",
                "nombre_obra_social": "Swiss Medical",
                "ids_alergias": [1],
                "ids_antecedentes": []
            }
        }
    )
    
    nombre: Optional[str] = None
    apellido: Optional[str] = None
    fecha_nacimiento: Optional[date] = None
    email: Optional[str] = None
    telefono: Optional[str] = None
    nombre_grupo_sanguineo: Optional[GrupoSanguineoType] = None
    nombre_obra_social: Optional[str] = None
    ids_alergias: Optional[list[int]] = None
    ids_antecedentes: Optional[list[int]] = None

class PacienteRead(PacienteBase):
    alergias: list[AlergiaRead] = []
    antecedentes: list[AntecedenteRead] = []
    obra_social: Optional[ObraSocialRead] = None
    grupo_sanguineo: Optional[GrupoSanguineoRead] = None

File: frontend/turnero-medico/src/types/Paciente.ts
// turnero-medico/src/types/Paciente.ts

export interface Alergia {
  id: number;
  nombre: string;
}

export interface Antecedente {
  id: number;
  descripcion: string;
}

export interface GrupoSanguineo {
  nombre: string;
}

export interface ObraSocialMini {
  nombre: string;
  porcentaje_cobertura: number;
  nombre_tipo: string;
}

export interface TurnoPaciente {
  id: number;
  fecha: string;
  horario: string;
  consultorio: string;
}

export interface Paciente {
  dni: number;
  nombre: string;
  apellido: string;
  fecha_nacimiento: string;
  email: string;
  telefono: string;

  nombre_grupo_sanguineo: string;
  nombre_obra_social?: string | null;

  grupo_sanguineo: GrupoSanguineo;
  obra_social?: ObraSocialMini | null;

  alergias: Alergia[];
  antecedentes: Antecedente[];
  turnos: TurnoPaciente[];
}

/* ===========================
      CREATE
=========================== */

export interface PacienteCreate {
  dni: number;
  nombre: string;
  apellido: string;
  fecha_nacimiento: string;
  email: string;
  telefono: string;

  nombre_grupo_sanguineo: string;
  nombre_obra_social?: string | null;

  ids_alergias?: number[];
  ids_antecedentes?: number[];
}

/* ===========================
      UPDATE
=========================== */

export interface PacienteUpdate {
  nombre?: string;
  apellido?: string;
  fecha_nacimiento?: string;
  email?: string;
  telefono?: string;

  nombre_grupo_sanguineo?: string | null;
  nombre_obra_social?: string | null;

  ids_alergias?: number[] | null;
  ids_antecedentes?: number[] | null;
}
